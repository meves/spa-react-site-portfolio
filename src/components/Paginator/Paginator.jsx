import React from "react";
import style from './Paginator.module.scss'; 
import { useState } from 'react';

const Paginator = (props) => {
    const [currentBlock, setCurrentBlock] = useState(1);
    const left = (currentBlock - 1) * 10 + 1;
    const right = currentBlock * 10;
    const blockSize = 10;

    let numberOfPages = Math.ceil(props.totalCount / props.count);
    let numberOfBlocks = Math.ceil(numberOfPages / blockSize);
    
    const blocks = [];
    for (let i = left; i <= right && i <= numberOfPages; i++) {
        const classes = `${style.paginationNumber} ` + 
            `${i === props.currentPage && style.currentNumber}`;
            blocks.push(<div className={classes} key={i}
                        onClick={() => {
                            props.handlePageNumberClick(i);
                        }}
                    >{i}</div>);
    }
    return (
        <div className={style.paginationWrapper}>
            {currentBlock > 1 ? <button onClick={() => setCurrentBlock(1)}>First</button> : null}
            {currentBlock > 1 ? <button onClick={() => setCurrentBlock(currentBlock-1)}>{`Prev <<`}</button> : null}            
            <div className={style.pagesWrapper}>
                { blocks }
            </div>
            {currentBlock < numberOfBlocks ? <button onClick={() => setCurrentBlock(currentBlock+1)}>{`>> Next`}</button> : null}
            {currentBlock < numberOfBlocks ? <button onClick={() => setCurrentBlock(numberOfBlocks)}>Last</button> : null}
        </div>
    );          
}

export default Paginator;
