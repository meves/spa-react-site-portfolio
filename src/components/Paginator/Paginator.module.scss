.paginationWrapper {
    display: flex;
}

.pagesWrapper {
    display: flex;
    justify-content: space-around;
    width: 25em;
}

.paginationNumber {
    flex: 1;
    margin-right: 0.5em;
    font-size: 0.75rem;
}
.paginationNumber:hover {
    cursor: pointer;
    color: var(--bg-color);
}
.paginationNumber:first-child {
    margin-left: 0.4em;
}

.currentNumber {
    font-weight: bold;
    text-decoration: underline;
    color: red;
}
