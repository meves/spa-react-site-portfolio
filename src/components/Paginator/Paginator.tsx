import React, { FC } from "react";
import styles from './Paginator.module.scss'; 
import { useState } from 'react';
import classNames from "classnames";

type PropsType = {
    count: number
    totalCount: number
    currentPage: number
    handlePageNumberClick: (currentPage: number) => void
}

const Paginator: FC<PropsType> = (props): JSX.Element => {
    const [currentBlock, setCurrentBlock] = useState(1);
    const left: number = (currentBlock - 1) * 10 + 1;
    const right: number = currentBlock * 10;
    const blockSize: number = 10;

    let numberOfPages: number = Math.ceil(props.totalCount / props.count);
    let numberOfBlocks: number = Math.ceil(numberOfPages / blockSize);
    
    const blocks: Array<JSX.Element> = [];
    for (let i: number = left; i <= right && i <= numberOfPages; i++) {
        const classes: string = classNames(styles.paginationNumber, {
                                  [styles.currentNumber]: i === props.currentPage
                        });            
        blocks.push(<div className={classes} key={i}
                         onClick={() => {props.handlePageNumberClick(i);}}>{i}
                    </div>
        );
    }
    return (
        <div className={styles.paginationWrapper}>
            {currentBlock > 1 ? <button onClick={() => setCurrentBlock(1)}>First</button> : null}
            {currentBlock > 1 ? <button onClick={() => setCurrentBlock(currentBlock-1)}>{`Prev <<`}</button> : null}            
            <div className={styles.pagesWrapper}>
                { blocks }
            </div>
            {currentBlock < numberOfBlocks ? <button onClick={() => setCurrentBlock(currentBlock+1)}>{`>> Next`}</button> : null}
            {currentBlock < numberOfBlocks ? <button onClick={() => setCurrentBlock(numberOfBlocks)}>Last</button> : null}
        </div>
    );          
}

export default Paginator;
