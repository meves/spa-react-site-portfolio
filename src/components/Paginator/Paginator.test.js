import React from "react";
//import { act } from "react-dom/test-utils";
import { create, act } from 'react-test-renderer';
import Paginator from "./Paginator";

describe('Paginator component', () => {
   test('renders correctly', () => {
       const component = create(<Paginator totalCount={1000} count={5} currentPage={1}/>);
       const tree = component.toJSON();
       expect(tree).toMatchSnapshot();
   });
});
