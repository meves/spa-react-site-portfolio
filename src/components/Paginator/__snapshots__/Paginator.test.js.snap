// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator component renders correctly 1`] = `
<div
  className="paginationWrapper"
>
  <div
    className="pagesWrapper"
  >
    <div
      className="paginationNumber currentNumber"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      2
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      3
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      5
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      6
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      7
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      8
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      9
    </div>
    <div
      className="paginationNumber false"
      onClick={[Function]}
    >
      10
    </div>
  </div>
  <button
    onClick={[Function]}
  >
    &gt;&gt; Next
  </button>
  <button
    onClick={[Function]}
  >
    Last
  </button>
</div>
`;
